blogposts/2021/how-we-built-batch-changes.md

```diff
@@ -0,0 +1,195 @@
+tags: [blog]
+slug: how-we-built-batch-changes
+<!-- TODO: heroImage: /blog/TODO.jpg -->
+<!-- TODO: socialImage: https://about.sourcegraph.com/blog/TODO.jpg -->
+published: false
+
+---
+
+Straightforward is the last word I’d use to describe how we built Batch Changes.
+Exciting, rewarding, surprising, iterative, customer-focused - one of those
+maybe, but not straightfoward.
+
+Why? The two name changes in 1.5 years of development - from Automation to
+Campaigns to Batch Changes - give you a glimpse. The real story, though, the one
+that makes me doubt whether we actually pulled this off, I think is best told by
+five pull requests.
+
+Not because these pull requests solve hard technical problems in two lines of
+clever code or fix nasty bugs that held us back for months. It's because they
+show how we avoided two big dangers of iteration-based product development:
+local maximums and the sunken cost fallacy.
+
+To give you some context for these terms, let me tell you how I see the premise
+of iteration-based product development. It doesn't matter whether you call it
+_Agile_, _Sprint, _Shape Up_ or something else, the idea is this: until you've
+built the product, you don't _really_ know what it should look like or how to
+best build it. You've never used it, so how would you know how it best works?
+You also haven't built it (yet), so you don't have experience building it.
+Chances that you can accurately predict how its development will go are low.
+
+Acknowledging that you _don't_ plan the development of the product from start to
+finish in every detail. You _don't_ nail down and define the hundreds of steps
+it will take to go from 0 to fully-built and shipped and launched product.
+Instead, working iteration-based, you define roughly where you want to go and
+then take a single small step. Then you look around, get your bearings, correct
+course if necessary, and take another step. Build something, ship it, learn,
+build again.
+
+I personally love working like this and think that in a lot of contexts in which
+software is being developed (fast-moving startups, for example) it's the best way to
+ship successful products.
+
+But there's one huge problem with this approach: it's really easy to iterate
+yourself into a corner. If all you do is take small steps you can miss the point
+where you should take big leaps to someplace else entirely. You end up with [a
+local maximum](https://elezea.com/2012/09/iphone-5-local-maximum/). On top of
+that: even if you realise that you're stuck and _know_ that you must take a big
+leap, it's hard to abandon the things you've built in many, many small
+iterations. "So much work went into this! We can't simply delete it! Right?"
+Right, it's the [sunken cost
+fallacy](https://en.wikipedia.org/wiki/Sunk_cost#Fallacy_effect).
+
+Back to the five pull requests and how we've managed to defy both of these dangers.
+
+In September 2019 we started working on Batch Changes, except that back then it
```

Comment:
```suggestion
In September 2019 after Eric Brody-Moore had the original idea we started working on Batch Changes, except that back then it
```

Reply:
```
In September 2019, _despite_ Eric Brody-Moore, we started working on Batch Changes, except that back then it
```

Reply:
Hmm it seems like you're trying to re-write history here 🤔

Reply:
That's _exactly_ what I'm trying to do, yes!